ksqldb-engine/src/main/java/io/confluent/ksql/engine/TransientQueryCleanupService.java

```diff
@@ -0,0 +1,210 @@
+  private final KafkaTopicClient topicClient;
+  private final int initialDelay;
+  private final int intervalPeriod;
+  private Optional<Set<String>> localCommandsQueryAppIds;
+  private QueryRegistry queryRegistry;
+  private int numLeakedTopics;
+  private int numLeakedStateFiles;
+
+  public TransientQueryCleanupService(final ServiceContext serviceContext,
+                                      final KsqlConfig ksqlConfig) {
+    final String internalTopicPrefix = buildInternalTopicPrefix(ksqlConfig, false);
+    this.internalTopicPrefixPattern = Pattern.compile(internalTopicPrefix);
+    this.transientAppIdPattern = Pattern.compile(internalTopicPrefix + ".*_[0-9]\\d*_[0-9]\\d*");
+
+    this.initialDelay = ksqlConfig.getInt(
+            KsqlConfig.KSQL_TRANSIENT_QUERY_CLEANUP_SERVICE_INITIAL_DELAY_SECONDS);
+
+    this.intervalPeriod = ksqlConfig.getInt(
+            KsqlConfig.KSQL_TRANSIENT_QUERY_CLEANUP_SERVICE_PERIOD_SECONDS);
+
+    this.stateDir = ksqlConfig.getKsqlStreamConfigProps()
+            .getOrDefault(
+                    StreamsConfig.STATE_DIR_CONFIG,
+                    StreamsConfig.configDef()
+                            .defaultValues()
+                            .get(StreamsConfig.STATE_DIR_CONFIG))
+            .toString();
+
+    this.topicClient = serviceContext.getTopicClient();
+    this.localCommandsQueryAppIds = Optional.empty();
+    this.queriesGuaranteedToBeRunningAtSomePoint = new HashSet<>();
+    this.numLeakedTopics = 0;
+    this.numLeakedStateFiles = 0;
+  }
+
+  @Override
+  protected void runOneIteration() {
+    try {
+      final List<String> leakedTopics = findLeakedTopics();
+      this.numLeakedTopics = leakedTopics.size();
+      LOG.info("Cleaning up {} leaked topics: {}", numLeakedTopics, leakedTopics);
+      getTopicClient().deleteTopics(leakedTopics);
+    } catch (Throwable t) {
+      LOG.error(
+          "Failed to clean up topics with exception: " + t.getMessage(), t);
+    }
+    try {
+      final List<String> leakedStateDirs = findLeakedStateDirs();
+      this.numLeakedStateFiles = leakedStateDirs.size();
+      LOG.info("Cleaning up {} leaked state directories: {}",
+          numLeakedStateFiles,
+          leakedStateDirs.stream().map(file -> stateDir + "/" + file)
+                  .collect(Collectors.toList()));
+      leakedStateDirs.forEach(this::deleteLeakedStateDir);
+    } catch (Throwable t) {
+      LOG.error(
+          "Failed to clean up state directories with exception: " + t.getMessage(), t);
+    }
+  }
+
+  @Override
+  public Scheduler scheduler() {
+    return Scheduler.newFixedRateSchedule(initialDelay, intervalPeriod, TimeUnit.SECONDS);
+  }
+
+  public void setQueryRegistry(final QueryRegistry queryRegistry) {
+    this.queryRegistry = queryRegistry;
+  }
+
+  @SuppressFBWarnings(value = "EI_EXPOSE_REP2")
+  public void setLocalCommandsQueryAppIds(final Set<String> ids) {
+    this.localCommandsQueryAppIds = Optional.of(ids);
```

Comment:
Maybe leave a comment about how this is is set by the startup thread and therefore doesn't need synchronization since it's set before this service is started.

Reply:
will take care of it in the follow-up